docs/modules/sql/pages/user-defined-types.adoc

```diff
@@ -0,0 +1,630 @@
+more complex cases of fully/partially connected graphs of objects. See <<queryingExamples,examples>> for more information.
+
+WARNING: Support for UDTs is an experimental feature, and disabled by default.
+See <<enabling-udt-support, Enabling UDT support>> to learn how to enable it.
+
+== Feature Overview
+
+Due to experimental nature of UDTs, feature set is limited to the following:
+
+- Support for three formats - `portable`, `compact` and `java` with varying level of sub-feature support
+- Instance and type-level cycles are only supported for Java types.
+- Type formats (referred to as type kind or kind hereafter) can not be mixed, both between mappings and types themselves.
+For example, you cannot use Java type in another Portable type or Portable mapping.
+- Limited support for instance and type-level cycles; only for Java types.
+Note that INSERT/UPDATE are disabled for mappings that use type hierarchies that contain cycles.
+- Limited support for `INSERT` and `UPDATE` queries for mappings using UDTs; if a `TYPE` hierarchy contains cycles,
+`INSERT` s and `UPDATE` s become disabled for the mapping that uses this `TYPE` hierarchy.
+- Support for using UDT-based projections in both normal `SELECT` projection lists, `WHERE` filters and in `JOIN` conditions.
+
+== Enabling UDT Support
+You can enable the UDT usage by setting the `hazelcast.sql.experimental.custom.types.enabled` property to `true` in the configuration.
+[tabs]
+==== 
+XML:: 
++ 
+-- 
+[source,xml]
+----
+<hazelcast>
+    <properties>
+        <property name="hazelcast.sql.experimental.custom.types.enabled">true</property>
+    </properties>
+</hazelcast>
+----
+--
+
+YAML::
++
+[source,yaml]
+----
+hazelcast:
+  properties:
+    hazelcast.sql.experimental.custom.types.enabled: true
+----
+
+Java::
++
+[source,java]
+----
+final Config config = new Config();
+config.setProperty("hazelcast.sql.experimental.custom.types.enabled", "true");
+----
+====
+
+== Creating Types
+
+Types can be created with the `CREATE TYPE` command:
+
+[source,sql]
+----
+CREATE [OR REPLACE] TYPE [IF NOT EXISTS] MyTypeName
+    [(colName colType, ...)]
+    OPTIONS (
+    'format'='{java|portable|compact}'
+    [, 'javaClass'='com.myPackage.MyJavaClass'] <1>
+    [, 'compactTypeName'='MyCompactRecordTypeName'] <2>
+    [, 'portableFactoryId'='123', 'portableClassId'='456', ['portableVersion'='789']] <3>
+    )
+----
+<1> `java` format requires the `javaClass` option.
+<2> `portable` format requires `portableFactoryId` and `portableClassId` and optionally `portableVersion`.
+<3> `compact` format requires `compactTypeName` - this is not the name of the created type, but rather internal name of the Compact record type, used internally by the Compact Serialization format.
+
+NOTE: The column list can be omitted in some cases. However, to reference another Type, it has to be explicitly specified in the column list.
```

Comment:
```suggestion
NOTE: To reference another type, you must provide the column list. Otherwise, the column may be automatically resolved as an `OBJECT` type.
```